### PR TITLE
symbolic backend

### PR DESCRIPTION
**Pull request by jorgefandinno**
- adding examples
- alternative way of using the SymbolicBackend
- complete tests for symbolic backend

### CODE REVIEW

File: clingox/backends.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
+'''
+This module provides a backedn wrapper to work with symbols instead of integer literals.
+'''
+from typing import Iterable, Tuple
+from itertools import chain
+from clingo import HeuristicType, Symbol, Backend, TruthValue
+
+# pylint: disable=line-too-long
```
Comment (rkaminsk) [inline clingox/backends.py]: I already put the line length limit quite high. In the file it is mostly the comments that go beyond the limit. With my editor it is rather easy to auto textwrap the comments. If you want to, I can do it quickly. We should not disable it.

File: clingox/backends.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        '''
+        condition = chain(self._add_symbols(pos_condition), self._add_negated_symbols(neg_condition))
+        return self.backend.add_acyc_edge(node_u, node_v, condition)
+
+    # TODO: can this get negative literals?
+    def add_assume(self, literals: Iterable[Symbol]) -> None:
+        '''
+        Add assumptions to the underline backend.
+
+        :param literals: The list of symbols to assume true.
```
Comment (rkaminsk) [inline clingox/backends.py]: We should not switch between docstring styles. Better stick to numpy style comments.

File: clingox/backends.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+            An unsigned integer.
+        pos_condition : Iterable[Symbol]
+            List of program literals forming the positive part of the condition.
+        neg_condition : Iterable[Symbol]
+            List of program literals forming the negated part of the condition.
+
+        Returns
+        -------
+        None
+    '''
```
Comment (rkaminsk) [inline clingox/backends.py]: Misplaced '''

File: clingox/backends.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        Notes
+        -----
+        Follows Python's __exit__ conventions. Does not suppress exceptions.
+        '''
+        return self.backend.__exit__(type_, value, traceback)
+
+    # TODO: are node_u and node_v literals? do they need to be Symbols?
+    def add_acyc_edge(self, node_u: int, node_v: int, pos_condition: Iterable[Symbol], neg_condition: Iterable[Symbol]) -> None:
+        '''
+        Add an edge directive to the underline backend.
```
Comment (rkaminsk) [inline clingox/backends.py]: You write "underline backend" here but it should be "underlying backend". This happens also a lot elsewhere.

File: clingox/backends.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        neg_condition : Iterable[Symbol]
+            List of program symbols forming negated part of the condition
+        Returns
+        -------
+        None
+        '''
+        condition = chain(self._add_symbols(pos_condition), self._add_negated_symbols(neg_condition))
+        return self.backend.add_acyc_edge(node_u, node_v, condition)
+
+    # TODO: can this get negative literals?
```
Comment (rkaminsk) [inline clingox/backends.py]: Yes. Positive literals are assumed true and negative literals false.

File: clingox/backends.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+            List of program symbols to project on.
+
+        Returns
+        -------
+        None
+        '''
+        atoms = (self._add_symbols(symbols))
+        return self.backend.add_project(atoms)
+
+    # TODO: can the head also have negative literals?
```
Comment (rkaminsk) [inline clingox/backends.py]: No, all literals have to be positive.

File: clingox/backends.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+            The symbols associated with the atoms.
+
+        Returns
+        -------
+        Iterable[Tuple[int,int]]
+            The pairs representing the weighted atoms.
+        '''
+        return ((self.backend.add_atom(symbol), w) for (symbol, w) in weighted_symbols)
+
+    def _add_negated_symbols_weights(self, weighted_symbols: Iterable[Tuple[Symbol, int]]):
```
Comment (rkaminsk) [inline clingox/backends.py]: I would solve such things with a flag and just have one function to keep the code compact. Also you should spell check your docstrings.

File: clingox/backends.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+            The symbols associated with the atoms.
+
+        Returns
+        -------
+        Iterable[int]
+            The program atoms representing the atoms.
+        '''
+        return (self.backend.add_atom(symbol) for symbol in symbols)
+
+    def _add_negated_symbols(self, symbols: Iterable[Symbol]):
```
Comment (rkaminsk) [inline clingox/backends.py]: I would combine with the previous function to keep the code compact. `_add_symbols(self, sign: bool, symbols: Iterable[Symbol])` should do.

**Review by rkaminsk — CHANGES_REQUESTED** [review]

**Comment by jorgefandinno** [comment]
Please go ahead

**Pull request closed without merging**